lib/GeometryUtilities.js

```diff
@@ -34,227 +17,53 @@ const getLongestEdgeIndex = bb => {
-const intersectTri = ( mesh, geo, raycaster, ray, tri, intersections, seenFaces ) => {
-
-	const faceIndex = tri;
-	if ( seenFaces != null && seenFaces[ faceIndex ] ) {
-
-		return null;
+const intersectTri = ( mesh, geo, raycaster, ray, tri, intersections ) => {
 
-	}
-	if ( geo.isBufferGeometry ) {
-
-		tri = tri * 3;
-		const a = geo.index ? geo.index.getX( tri ) : tri;
-		const b = geo.index ? geo.index.getX( tri + 1 ) : tri + 1;
-		const c = geo.index ? geo.index.getX( tri + 2 ) : tri + 2;
-
-		const intersection = checkBufferGeometryIntersection( mesh, raycaster, ray, geo.attributes.position, geo.attributes.uv, a, b, c );
+	const triOffset = tri * 3;
+	const a = geo.index ? geo.index.getX( triOffset ) : triOffset;
+	const b = geo.index ? geo.index.getX( triOffset + 1 ) : triOffset + 1;
+	const c = geo.index ? geo.index.getX( triOffset + 2 ) : triOffset + 2;
 
-		if ( intersection ) {
+	const intersection = checkBufferGeometryIntersection( mesh, raycaster, ray, geo.attributes.position, geo.attributes.uv, a, b, c );
 
-			if ( seenFaces != null ) {
+	if ( intersection ) {
 
-				seenFaces[ faceIndex ] = true;
-
-			}
-			intersection.faceIndex = faceIndex; // triangle number
-			if ( intersections ) intersections.push( intersection );
-			return intersection;
-
-		}
-
-	} else if ( geo.isGeometry ) {
-
-		const faces = geo.faces;
-		const vertices = geo.vertices;
-		const uvs = geo.uvs;
-		const face = faces[ tri ];
-		const isMultiMaterial = Array.isArray( mesh.material );
-		const faceMaterial = isMultiMaterial ? mesh.material[ face.materialIndex ] : mesh.material;
-
-		const uvA = new THREE.Vector2();
-		const uvB = new THREE.Vector2();
-		const uvC = new THREE.Vector2();
-
-		if ( faceMaterial !== undefined ) {
-
-			const fvA = vertices[ face.a ];
-			const fvB = vertices[ face.b ];
-			const fvC = vertices[ face.c ];
-
-			const intersection = checkIntersection( mesh, faceMaterial, raycaster, ray, fvA, fvB, fvC, intersectionPoint );
-
-			if ( intersection ) {
-
-				if ( seenFaces != null ) {
-
-					seenFaces[ faceIndex ] = true;
-
-				}
-
-				if ( uvs && uvs[ faceIndex ] ) {
-
-					const uvsf = uvs[ faceIndex ];
-					uvA.copy( uvsf[ 0 ] );
-					uvB.copy( uvsf[ 1 ] );
-					uvC.copy( uvsf[ 2 ] );
-
-					intersection.uv = uvIntersection( intersectionPoint, fvA, fvB, fvC, uvA, uvB, uvC );
-
-				}
-
-				intersection.face = face;
-				intersection.faceIndex = tri;
-				if ( intersections ) intersections.push( intersection );
-				return intersection;
-
-			}
-
-		}
+		intersection.faceIndex = tri;
+		if ( intersections ) intersections.push( intersection );
+		return intersection;
 
 	}
+
 	return null;
 
 };
 
-const intersectTris = ( mesh, geo, raycaster, ray, tris, intersections, seenFaces ) => {
+const intersectTris = ( mesh, geo, raycaster, ray, offset, count, intersections ) => {
 
-	for ( let i = 0, l = tris.length; i < l; i ++ ) {
+	for ( let i = offset; i < offset + count; i ++ ) {
```

Comment:
Caching the length is an optimization for loops over long arrays so I think we should keep that here -- it's used pretty consistently in the THREE.js codebase, as well. Here's a quick test I ran in chrome with the results:

```js
var arr = new Array(10000000);
console.time('uncached'); for(let i = 0; i < arr.length; i ++); console.timeEnd('uncached');
console.time('cached'); for(let i = 0, l = arr.length; i < l; i ++); console.timeEnd('cached');
// uncached: 54.7998046875ms
// cached: 5.06689453125ms
```

It might be worth adding it in a few other places, as well.

Reply:
Interesting, I guess I got sloppy because to be honest I doubted it would matter with modern JS interpreters. I will fix it up.